kernel/aster-nix/src/process/process_vm/init_stack/mod.rs

```diff
@@ -349,30 +371,32 @@ fn generate_random_for_aux_vec() -> [u8; 16] {
 /// A reader to parse the content of an `InitStack`.
 pub struct InitStackReader<'a> {
     base: Vaddr,
-    vmar: &'a Vmar<Full>,
+    vmo: &'a Vmo,
+    /// The mapping address of the `InitStack`.
+    map_addr: usize,
 }
 
 impl<'a> InitStackReader<'a> {
     /// Read argc from the process init stack
     pub fn argc(&self) -> Result<u64> {
         let stack_base = self.user_stack_top();
-        Ok(self.vmar.read_val(stack_base)?)
+        Ok(self.vmo.read_val(stack_base - self.map_addr)?)
     }
 
     /// Read argv from the process init stack
     pub fn argv(&self) -> Result<Vec<CString>> {
         let argc = self.argc()? as usize;
-        // base = stack bottom + the size of argc
-        let base = self.user_stack_top() + 8;
+        // vmo_base = stack bottom + the size of argc
+        let vmo_base = self.user_stack_top() + 8 - self.map_addr;
```

Comment:
So the documentation of the `user_stack_top` function is truely misleading. I suggest removing it and use `self.pos()` here. It is also not "stack bottom". The absolute understandable version should be:

```rust
// The offset of the inital stack position in Vmo is:
// initial stack position + the size of `argc` in memory - Vmo mapping offset
let offset_of_pos_in_vmo = self.pos() + 8 - self.map_addr
```

Reply:
The `user_stack_top` here is the method in `InitStackReader` instead of the one in `InitStack`. After initialization, the bottom of the init stack is equal to the top of the user stack. Current documents is indeed ambiguous and I will refine them.